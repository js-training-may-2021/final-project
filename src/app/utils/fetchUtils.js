import axios from 'axios';
import _ from 'lodash';

const isDev = process.env.NODE_ENV === 'development';
const pathApi = isDev ? 'http://localhost:3000' : '/api/v1';
export const limit = 20;

const fetchPokemons = (page, query, attr, order) => axios.get(`${pathApi}/pokemons?q=${query}&_page=${page}&_limit=${limit}&_sort=${attr}&_order=${order}`);
const normalizedQuery = (query) => _.trimStart(query, '#');

export const mappingFetchPokemons = {
  idAsc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'id', 'asc'),
  idDesc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'id', 'desc'),
  nameAsc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'name', 'asc'),
  nameDesc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'name', 'desc'),
};

export const fetchPokemonById = (id) => axios.get(`${pathApi}/pokemons?id=${id}`);

export const fetchData = () => (
  axios({
    method: 'post',
    url: '/api/v1/data',
    data: { token: localStorage.token },
    timeout: 4000,
  })
);

export const fetchCatch = (pokemon) => axios({
  method: 'post',
  url: '/api/v1/catchPokemon',
  data: { ...pokemon, token: localStorage.token },
  timeout: 4000,
});

export const fetchLogin = (data) => axios({
  method: 'post',
  url: '/api/v1/login',
  data,
  timeout: 4000,
});
