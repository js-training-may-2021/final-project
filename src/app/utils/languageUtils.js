import { useTranslation } from 'react-i18next';

export const setDefaultLanguage = () => {
  const defaultLanguage = 'ru';

  if (localStorage.getItem('language') === null) {
    localStorage.setItem('language', defaultLanguage);
  }
};

export const getCurrentLanguage = () => localStorage.getItem('language');

export const toggleLanguage = () => {
  const currentLanguage = localStorage.getItem('language');
  const newLanguage = currentLanguage === 'ru' ? 'en' : 'ru';
  localStorage.setItem('language', newLanguage);

  return newLanguage;
};

export const loc = (text) => {
  const { t } = useTranslation();
  return t(text);
};
