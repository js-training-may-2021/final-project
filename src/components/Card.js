import React from 'react';
import Button from '../containers/Button';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
const List = styled.ul`
    margin: 0 auto;
    text-align: center;
`

const Element = styled.li`
    display: inline-block;
    width: 30%;
    max-width: 300px;
    background-color: #fff;
    border-radius: 16px;
    margin-right: 30px;
    margin-bottom: 60px;
    color: #000;
    box-shadow: 0px 4px 30px rgba(0, 0, 0, 0.07);

    @media (max-width: 719px) {
        width: 300px;
        display: block;
        margin: 0 auto;
        margin-bottom: 25px;
    }
    @media (max-width: 370px) {
        width: 270px;
    }
`

const Photo = styled.img`
    max-width: 200px;
    max-height: 200px;
    @media (max-width: 370px) {
        max-width: 170px;
    }
`
const Info = styled.div`
    @media (max-width: 719px) {
        padding: 20px 20px 0;
    }
    @media (max-width: 719px) {
        padding: 10px 10px 0;
    }
`

const Name = styled.h4`
    font-size: 26px;
    line-height: 36px;

`

const Text = styled.p`
    font-size: 13px;
    line-height: 19px;
    color: #C1C1C1;
    margin-bottom: 7px;
`


const Card = ({pokemons}) => {

    return (
        <List>
            {
                pokemons.map(pokemon => {
                    const isCaught = pokemon.caught ? true : false;
                    return (
                        <Link to={`/pokemons/${pokemon.id}`} key={pokemon.id}>
                            <Element >
                                <Info>
                                  {(pokemon.id<721) ? <Photo src={require(`../assets/pokemons/${pokemon.id}.png`).default} alt='' /> : null}
                                    <Name>{pokemon.name.charAt(0).toUpperCase()+pokemon.name.slice(1)}</Name>
                                    <Text>{'ID: ' + pokemon.id}</Text>  
                                </Info>
                                <Button disabled={isCaught} id={pokemon.id} name={pokemon.name} isCaught={isCaught}/>
                            </Element> 
                        </Link>
                )})  
            }
        </List>
    )
}

export default Card
