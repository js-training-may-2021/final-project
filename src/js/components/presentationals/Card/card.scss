@import '../../../../styles/variables';

.card {
  width: 240px;
  margin: 20px 0;
  padding: 20px;
  display: flex;
  flex-direction: column;

	&--caught {
		opacity: 0.5;
	}

  &__title {
    color: map-get($colorScheme, "text");
    word-wrap: break-word;
  }
}