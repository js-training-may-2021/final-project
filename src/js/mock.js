export const pokemons = [
  {
    name: 'bulbasaur',
    id: 1,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'ivysaur',
    id: 2,
    isCaught: true,
    catchDate: '23 June 2021',
  },
  {
    name: 'venusaur',
    id: 3,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'charmander',
    id: 4,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'charmeleon',
    id: 5,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'charizard',
    id: 6,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'squirtle',
    id: 7,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'wartortle',
    id: 8,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'blastoise',
    id: 9,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'caterpie',
    id: 10,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'metapod',
    id: 11,
    isCaught: false,
    catchDate: null,
  }];  

export const uncaughtPokemon = pokemons[0];
export const caughtPokemon = pokemons[1];
