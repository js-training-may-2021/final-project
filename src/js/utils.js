import {createMemoryHistory} from 'history';
import configureStore from 'redux-mock-store';

export const POKEMONS_PER_PAGE = 10;
const START_PAGE = 1;

let currentPage = START_PAGE;

export const getCurrentPage = () => {
	return currentPage++;
};

export const AppRoute = {
  MAIN: '/',
  DETAIL: '/detail/:id',
  CAUGHT: '/caught'
};

const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

export const getCatchDate = () => {
  const date = new Date();

  const day = date.getDate();
  const month = monthNames[date.getMonth()];
  const year = date.getFullYear();

  return `${day} ${month} ${year}`
};

export const noop = () => {};
export const history = createMemoryHistory();
const mockStore = configureStore();
export const testStore = mockStore({});
