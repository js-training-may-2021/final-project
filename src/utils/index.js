export const firstLetterToUpperCase = (str) => {
  if (str === undefined) {
    return str;
  } else {
    return str
      .toLowerCase()
      .split(" ")
      .map((s) => s.charAt(0).toUpperCase() + s.substring(1))
      .join(" ");
  }
};

export const formatDate = (dateInput) => {
  return dateInput.toLocaleString("ru-RU");
};

export const padToThree = (number) => number <= 999 ? `00${number}`.slice(-3) : number;

export const handleErrors = (response) => {
  if (!response.ok) {
    throw Error(`${response.status}: ${response.statusText}`);
  }
  return response;
};
