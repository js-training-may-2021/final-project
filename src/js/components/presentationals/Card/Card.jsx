import React from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';
import Button from '../Button/Button.jsx';
import withToggle from '../../hocs/withToggle/withToggle.jsx';
import './card.scss';

const ButtonWrapped = withToggle(Button);

const Card = (props) => {
  const {
    pokemon,
    isToggleChecked,
    onCardClick,
    onButtonClick,
    onToggleClick,
  } = props;

	const {id, name} = pokemon;

	return (
		<li className={"list__card card" + (isToggleChecked ? " card--caught" : "")} onClick={() => onCardClick(pokemon)}>
			<Link className="card__link-wrap" to={`../../../../detail/${id}`}>
				<img className="card__img" src={require(`../../../../images/${id < 721 ? id : `default`}.png`)} alt={name[0].toUpperCase() + name.slice(1)} />
        <h2 className="card__title">{name[0].toUpperCase() + name.slice(1)}</h2>
			</Link>
      <ButtonWrapped
        isToggleChecked={isToggleChecked}
        onClick={() => {
          onButtonClick(pokemon);
          onToggleClick();
        }}
			/>
		</li>
	);
};

Card.propTypes = {
	pokemon: PropTypes.shape({
		id: PropTypes.number.isRequired,
		name: PropTypes.string.isRequired,
		isCaught: PropTypes.bool.isRequired,
		catchDate: PropTypes.oneOfType([() => null, PropTypes.instanceOf(Date)]),
	}).isRequired,
  isToggleChecked: PropTypes.bool.isRequired,
  onCardClick: PropTypes.func.isRequired,
  onButtonClick: PropTypes.func.isRequired,
  onToggleClick: PropTypes.func,
}

export default React.memo(Card);
