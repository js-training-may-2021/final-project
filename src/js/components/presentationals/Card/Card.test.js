import React from 'react';
import renderer from 'react-test-renderer';
import {Router} from 'react-router-dom';
import Card from './Card.jsx';
import {noop, history} from '../../../utils.js';
import {uncaughtPokemon, caughtPokemon} from '../../../mock.js';

it(`Render Card of uncaught pokemon`, () => {
  const tree = renderer
    .create(
      <Router history={history}>
        <Card
          pokemon={uncaughtPokemon}
          isToggleChecked={uncaughtPokemon.isCaught}
          onCardClick={noop}
          onButtonClick={noop}
          onToggleClick={noop}
        />
      </Router>)
    .toJSON();

  expect(tree).toMatchSnapshot();
  expect(tree.props.className).toBe('list__card card');
});

it(`Render Card of caught pokemon`, () => {
  const tree = renderer
    .create(
      <Router history={history}>
        <Card
          pokemon={caughtPokemon}
          isToggleChecked={caughtPokemon.isCaught}
          onCardClick={noop}
          onButtonClick={noop}
          onToggleClick={noop}
        />
      </Router>)
    .toJSON();

  expect(tree).toMatchSnapshot();
  expect(tree.props.className).toBe('list__card card card--caught');
});
