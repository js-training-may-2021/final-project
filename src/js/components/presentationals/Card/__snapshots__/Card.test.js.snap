// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Card of caught pokemon 1`] = `
<li
  className="list__card card card--caught"
  onClick={[Function]}
>
  <a
    className="card__link-wrap"
    href="/detail/2"
    onClick={[Function]}
  >
    <img
      alt="Ivysaur"
      className="card__img"
      src="test-file-stub"
    />
    <h2
      className="card__title"
    >
      Ivysaur
    </h2>
  </a>
  <button
    className="card__button button button--checked"
    onClick={[Function]}
    type="button"
  >
    Release
  </button>
</li>
`;

exports[`Render Card of uncaught pokemon 1`] = `
<li
  className="list__card card"
  onClick={[Function]}
>
  <a
    className="card__link-wrap"
    href="/detail/1"
    onClick={[Function]}
  >
    <img
      alt="Bulbasaur"
      className="card__img"
      src="test-file-stub"
    />
    <h2
      className="card__title"
    >
      Bulbasaur
    </h2>
  </a>
  <button
    className="card__button button"
    onClick={[Function]}
    type="button"
  >
    Catch!
  </button>
</li>
`;
